### PR TITLE
transformer model pipelines - base and QA

### PR DESCRIPTION
**Pull request by bfineran**
base framework for huggingface-like `pipeline` objects for transformer model inference. pipelines will be able to pull down a default model from sparsezoo for the task based on the available CPU. Adapted from `huggingface/transformers` and `patil-suraj/onnx_transformers`. Currently supports DeepSparse and onnxruntime, planned support for pytorch models

Example usage:
```python
>>> from deepsparse.transformers import pipeline
>>> pipe = pipeline("question-answering")
>>> pipe(question="what is my name?", context="my name is bfineran")
{'score': 0.501025378704071, 'start': 11, 'end': 18, 'answer': 'bfineran'} 
```

this PR will be converted from draft after style and documentation are updated

### CODE REVIEW

File: src/deepsparse/transformers/pipelines.py [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+            return self.model.run(list(inputs.values()))
+        # TODO: torch
+        # with self.device_placement():
+        #         with torch.no_grad():
+        #             inputs = self.ensure_tensor_on_device(**inputs)
+        #             predictions = self.model(**inputs)[0].cpu()
+        # if return_tensors:
+        #     return predictions
+        # else:
+        #     return predictions.numpy()
```
Comment (mgoin) [inline src/deepsparse/transformers/pipelines.py:255]: is this a todo for this diff?

File: src/deepsparse/transformers/pipelines.py [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+        sess_options = SessionOptions()
+        if num_cores is not None:
+            sess_options.intra_op_num_threads = num_cores
+        sess_options.log_severity_level = 3
+        sess_options.graph_optimization_level = GraphOptimizationLevel.ORT_ENABLE_ALL
+
+        onnx_model = onnx.load(onnx_path)
+        model = InferenceSession(
+            onnx_model.SerializeToString(), sess_options=sess_options
+        )
```
Comment (mgoin) [inline src/deepsparse/transformers/pipelines.py:791]: is there a reason why we aren't passing in the `onnx_path` directly?

**Review by mgoin — COMMENTED** [review]
looks great, mostly minor nits here. are we going to add a requirements.txt here or add a new "target" to the setup.py?

File: src/deepsparse/transformers/pipelines.py [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+
+    Base class implementing pipelined operations.
+    Pipeline workflow is defined as a sequence of the following operations:
+
+        Input -> Tokenization -> Model Inference ->
+        Post-Processing (task dependent) -> Output
+
+    Pipeline supports running with the DeepSparse engine or onnxruntime.
+
+    Some pipeline, output large tensor object as nested-lists.
```
[the remark this answers is not in the document]
Comment (bfineran) [reply src/deepsparse/transformers/pipelines.py]: removing, leftover from HF.

File: src/deepsparse/transformers/pipelines.py (hunk 1 above)
Comment (bfineran) [reply src/deepsparse/transformers/pipelines.py:255]: leaving for future work if we add torch support

File: src/deepsparse/transformers/pipelines.py [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+        model = InferenceSession(
+            onnx_model.SerializeToString(), sess_options=sess_options
+        )
+
+    return model, input_names
+
+
+def _validate_transformers_import():
+    if transformers_import_error is not None:
+        raise transformers_import_error
```
[the remark this answers is not in the document]
Comment (bfineran) [reply src/deepsparse/transformers/pipelines.py]: good idea. https://github.com/neuralmagic/deepsparse/pull/127

**Review by markurtz — DISMISSED** [review]
LGTM

File: src/deepsparse/transformers/pipelines.py [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+    engine_type: str,
+    num_cores: Optional[int],
+    num_sockets: Optional[int],
+) -> Tuple[Union[Engine, InferenceSession], List[str]]:
+    onnx_path, input_names, _ = _overwrite_model_inputs(model_path)
+
+    if engine_type == DEEPSPARSE_ENGINE:
+        model = compile_model(
+            onnx_path, batch_size=1, num_cores=num_cores, num_sockets=num_sockets
+        )
```
Comment (mgoin) [inline src/deepsparse/transformers/pipelines.py:785]: max_length isn't piped into _overwrite_model_inputs yet

Comment (bfineran) [reply src/deepsparse/transformers/pipelines.py]: good catch. fixed and tested

File: src/deepsparse/transformers/pipelines.py [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+
+    def predict(self, X):
+        """
+        Scikit / Keras interface to transformers' pipelines.
+        This method will forward to __call__().
+        """
+        return self(X=X)
+
+    def _parse_and_tokenize(
+        self, *args, padding=True, add_special_tokens=True, **kwargs
```
Comment (spacemanidol) [inline src/deepsparse/transformers/pipelines.py:227]: Is there an engine reason for setting padding to true? IIRC if its set to false the batches will be dynamically padded to max size for that batch.

**Review by natuan — DISMISSED** [review]
Left two small change requests.

On the compatibility with sklearn pipeline, I could see another way to refactor the QuestionAnsweringPipeline (specifically its current __call__) into (a) a transformer (in the data transformation sense) that converts input examples into features and (b) a predictor that produces the answers. The former would implements `fit_transform` and ends around the call to `squad_convert_examples_to_features`; and the latter implements `predict` and does the rest of the function. This way we could chain the same transformer to two different predictors (one for DeepSparse and the other for ORT)---right now there seems not much benefit of following sklearn pipeline standard.

File: src/deepsparse/transformers/pipelines.py (hunk 6 above)
Comment (bfineran) [reply src/deepsparse/transformers/pipelines.py:227]: this comes from huggingface's source code for the base pipeline. the QA pipeline doesn't follow this path, and goes through `squad_convert_examples_to_features` instead

**Review by markurtz — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Comment by bfineran** [comment]
@natuan I agree with your point, we could definitely add that in if needed. the current structure is there to match the huggingface API behavior

**Review by spacemanidol — COMMENTED** [review]
looks good
